### PR TITLE
EigenSolversApplication

### PR DESCRIPTION
**Pull request by oberbichler**
Details in [README](https://github.com/KratosMultiphysics/Kratos/blob/feature-eigen_solvers/applications/eigen_solvers_application/README.md).

Tested with and without MKL under Windows 10 and Ubuntu.

Maybe it is not possible to include this in binary releases (see #1114) but on my Windows machine its up to 130x faster than SuperLU. So it would be nice to have it in the main branch.

Greetings

### CODE REVIEW

**Comment by loumalouomega** [comment]
130x?, incredible

**Comment by oberbichler** [comment]
Just from a quick benchmark: Structural analysis with ~60.000 dofs:

SuperLU Windows: 66sec
SuperLU Linux: 16
Eigen_SparseLU: 8
Eigen_PardisoLU: 1
Eigen_PardisoLDLT: 0.7
Eigen_PardisoLLT: 0.5

Maybe it depends a bit on the problem/machine

File: applications/eigen_solvers_application/custom_python/add_custom_utilities_to_python.cpp [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+#include "spaces/ublas_space.h"
+
+namespace Kratos
+{
+
+namespace Python
+{
+
+void AddCustomUtilitiesToPython()
```
Comment (philbucher) [inline applications/eigen_solvers_application/custom_python/add_custom_utilities_to_python.cpp]: I think you can remove the files `add_custom_utilities_to_python.h` and `add_custom_utilities_to_python.cpp`, they are not doing anything

File: kratos/python_scripts/new_linear_solver_factory.py [hunk 2]
```diff
@@ -68,6 +68,20 @@ def ConstructSolver(configuration):
     elif (solver_type == "ParallelMKLPardisoSolver"):
         import KratosMultiphysics.MKLSolversApplication
         linear_solver = KratosMultiphysics.MKLSolversApplication.ParallelMKLPardisoSolver(configuration)
+    
+    ################################## following solvers need importing the EigenSolversApplication
+    elif(solver_type == "Eigen_SparseLU"):
```
Comment (philbucher) [inline kratos/python_scripts/new_linear_solver_factory.py]: I know that this factory will be moved to c++ soon, but still the comment on using the [Style Guide](https://github.com/KratosMultiphysics/Kratos/wiki/Style-Guide#json-configuration-file) for the names of the solver

File: applications/eigen_solvers_application/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+	add_definitions( -DEIGEN_USE_MKL_ALL )
+else()
+	message( "MKLROOT not defined" )
+endif()
+
+## generate variables with the sources
+set( KRATOS_EIGENSOLVERS_APPLICATION_SOURCES     
+	${CMAKE_CURRENT_SOURCE_DIR}/eigen_solvers_application.cpp 
+	${CMAKE_CURRENT_SOURCE_DIR}/custom_python/eigen_solvers_python_application.cpp
+	${CMAKE_CURRENT_SOURCE_DIR}/custom_python/add_custom_solvers_to_python.cpp
```
Comment (philbucher) [inline applications/eigen_solvers_application/CMakeLists.txt]: I would rename these files to sth with `linear` in it to make it clear what kind of solver it is, e.g.  `add_custom_linear_solvers_to_python`

File: applications/eigen_solvers_application/EigenSolversApplication.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+#      /_____/_/\__, /\___/_/ /_/ SolversApplication
+#              /____/
+# 
+#   Author: Thomas Oberbichler
+#
+
+from KratosEigenSolversApplication import *
+application = KratosEigenSolversApplication()
+application_name = "KratosEigenSolversApplication"
+application_folder = "eigen_solvers_application"
```
Comment (philbucher) [inline applications/eigen_solvers_application/EigenSolversApplication.py]: I think the naming should be in UpperCamelCase

File: applications/eigen_solvers_application/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+	message( FATAL_ERROR "EIGEN_ROOT not defined" )
+endif()
+
+if( MSVC )
+	set( CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /Ox" )
+else()
+	set( CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O3 -m64 -L$ENV{MKLROOT}/lib/intel64 -Wl,--no-as-needed -lpthread -lm -ldl" )
+endif()
+
+add_definitions( -DNDEBUG -DEIGEN_DEFAULT_TO_ROW_MAJOR -DEIGEN_MPL2_ONLY )
```
Comment (philbucher) [inline applications/eigen_solvers_application/CMakeLists.txt]: Is this coupled with the BuildMode `FullDebug`?
I think with full debug you should either add or remove the DNDEBUG

**Review by philbucher — COMMENTED** [review]
Small comments

File: applications/eigen_solvers_application/EigenSolversApplication.py (hunk 4 above)
Comment (oberbichler) [reply applications/eigen_solvers_application/EigenSolversApplication.py]: From the Kratos wiki template/script

File: applications/eigen_solvers_application/custom_python/add_custom_utilities_to_python.cpp (hunk 1 above)
Comment (oberbichler) [reply applications/eigen_solvers_application/custom_python/add_custom_utilities_to_python.cpp]: They should be used for the upcoming eigenvalue algorithms

File: kratos/python_scripts/new_linear_solver_factory.py (hunk 2 above)
Comment (oberbichler) [reply kratos/python_scripts/new_linear_solver_factory.py]: Did not saw this but I think "**solver**_type": "Eigen_SparseLU**Solver**" is a bit redundant.

File: applications/eigen_solvers_application/CMakeLists.txt (hunk 3 above)
Comment (oberbichler) [reply applications/eigen_solvers_application/CMakeLists.txt]: Then also the directory should be renamed. I try to keep it simple.

File: applications/eigen_solvers_application/CMakeLists.txt (hunk 5 above)
Comment (oberbichler) [reply applications/eigen_solvers_application/CMakeLists.txt]: It was just to ensure that the optimization is turned on. If the flags are correctly set by Release/FullDebug I can remove it.

File: applications/eigen_solvers_application/custom_python/add_custom_utilities_to_python.cpp (hunk 1 above)
Comment (armingeiser) [reply applications/eigen_solvers_application/custom_python/add_custom_utilities_to_python.cpp]: the eigenvalue algorithm will be also implemented as a solver, so i think the "...utilities..." can be removed

File: applications/eigen_solvers_application/custom_python/add_custom_solvers_to_python.cpp [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+#include <boost/python.hpp>
+
+// Project includes
+#include "includes/define.h"
+#include "processes/process.h"
+#include "custom_python/add_custom_solvers_to_python.h"
+
+#include "spaces/ublas_space.h"
+#include "linear_solvers/linear_solver.h"
+#include "custom_solvers/EigenDirectSolver.h"
```
Comment (pooyan-dadvand) [inline applications/eigen_solvers_application/custom_python/add_custom_solvers_to_python.cpp]: eigen_direct_solver.h would be style compatible

**Comment by pooyan-dadvand** [comment]
I was looking to the Eigen these days and seems that they are serious problem with aligning if you don't use their allocator. I'm not sure if you are storing some dense Eigen matrix (or vector), if yes, you should take a look to their [wiki](https://eigen.tuxfamily.org/dox/group__DenseMatrixManipulation__Alignement.html).

File: applications/eigen_solvers_application/custom_python/add_custom_solvers_to_python.cpp (hunk 6 above)
Comment (oberbichler) [reply applications/eigen_solvers_application/custom_python/add_custom_solvers_to_python.cpp]: Also the class name?

**Comment by oberbichler** [comment]
@pooyan-dadvand In my code I only generate wrappers around the raw data from the ublas objects. Of course the SparseLUSolver code from Eigen might allocate memory. But until now I never had problems.

File: applications/eigen_solvers_application/EigenSolversApplication.py (hunk 4 above)
Comment (philbucher) [reply applications/eigen_solvers_application/EigenSolversApplication.py]: @roigcarlo what is the current rule to follow?

File: kratos/python_scripts/new_linear_solver_factory.py (hunk 2 above)
Comment (philbucher) [reply kratos/python_scripts/new_linear_solver_factory.py]: I meant `solver_type = eigen_sparse_lu`

File: applications/eigen_solvers_application/EigenSolversApplication.py (hunk 4 above)
Comment (roigcarlo) [reply applications/eigen_solvers_application/EigenSolversApplication.py]: Should be in UpperCamelCase if possible indeed. Actually we are just updating the wiki. The new generator does it automatically.

File: applications/eigen_solvers_application/custom_python/add_custom_solvers_to_python.cpp (hunk 6 above)
Comment (pooyan-dadvand) [reply applications/eigen_solvers_application/custom_python/add_custom_solvers_to_python.cpp]: The class name should be CamelCase: `EigenDirectSolver`

**Comment by pooyan-dadvand** [comment]
> @pooyan-dadvand In my code I only generate wrappers around the raw data from the ublas objects. Of course the SparseLUSolver code from Eigen might allocate memory. But until now I never had problems.

Perfect! So nothing to be worried about!

**Comment by oberbichler** [comment]
@KratosMultiphysics/technical-committee I made some changes to match the styleguides. Is it OK now?

**Comment by RiccardoRossi** [comment]
@roigcarlo, @KratosMultiphysics/technical-committee as a follow up of yesterday's discussion and following @oberbichler and @armingeiser comments, i tool a look at the intel simplified software license

it looks like WE CAN include MKL

https://software.intel.com/en-us/mkl/license-faq

which would enable us access to the pardiso solver even on windows. What do you guys think?

File: applications/EigenSolversApplication/eigen_solvers_application.h [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+	virtual void PrintInfo(std::ostream &rOStream) const
+	{
+		rOStream << Info();
+		PrintData(rOStream);
+	}
+
+	/// Print object's data.
+	virtual void PrintData(std::ostream &rOStream) const
+	{
+		KRATOS_WATCH("in my application");
```
Comment (RiccardoRossi) [inline applications/EigenSolversApplication/eigen_solvers_application.h:106]: i think this watch is a leftover

**Comment by oberbichler** [comment]
For me it would be enough to have the header-only solver in the release (already faster than SuperLU) and Pardiso as an option for the "self-build" version.

Of course it would be great if it is possible to integrate MKL also into the release. Not only for the solvers but also for the BLAS/LAPACK implementation.

**Comment by RiccardoRossi** [comment]
@oberbichler i have been thinking about this, and i am very positive about the contribution. That is i think it would be ok to merge it now. 

I only have a suggestion, which sort of goes against what i told in the last days.
Wouldn't it be better to add this CONDITIONALLY to the External Solvers Application?

for example so that if someone adds COMPILE_EIGEN_INTERFACE and COMPILE_EIGEN_MKL_INTERFACE than it is a part of the ExternalSolversApplication?
(but importantly, if you do not add it then it is not included)

the reason for which i am telling this is that wherever ExternalSolversApplication is included than the new solvers would be available.

File: applications/EigenSolversApplication/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+## define library Kratos which defines the basic python interface
+add_library( KratosEigenSolversApplication SHARED ${KRATOS_EIGENSOLVERS_APPLICATION_SOURCES} )
+target_link_libraries( KratosEigenSolversApplication KratosCore )
+
+if( DEFINED ENV{MKLROOT} )
+	target_link_libraries( KratosEigenSolversApplication mkl_rt )
+endif()
+	
+set_target_properties( KratosEigenSolversApplication PROPERTIES PREFIX "" )
+install( TARGETS KratosEigenSolversApplication DESTINATION libs )
```
Comment (RiccardoRossi) [inline applications/EigenSolversApplication/CMakeLists.txt:56]: add
   
     cotire( KratosEigenSolversApplication )

File: applications/ExternalSolversApplication/custom_python/add_linear_solvers_to_python.cpp [hunk 9]
```diff
@@ -113,7 +113,6 @@ void  AddLinearSolversToPython()
     .def(init<double, unsigned int,  PreconditionerType::Pointer>())
```
Comment (RiccardoRossi) [inline applications/ExternalSolversApplication/custom_python/add_linear_solvers_to_python.cpp:113]: you should add a constructor which accepts a Parameter

**Review by RiccardoRossi — APPROVED** [review]

**Comment by oberbichler** [comment]
Thanks!

I also started by extending the `ExternalSolverApplication`. The only problem was, that the BLAS/LAPACK routines would be defined twice (libblas/liblapack + MKL). So I decided to create a seperate application to avoid this conflict. This was just the fastet solution at that moment (and I become more familiar with the application system of Kratos...).

Of course it would be nice to have only one "SolverApplication". We should definitly check if it is possibile to combine them.

**Comment by pooyan-dadvand** [comment]
For me the reason to be in a separate MKL application was the licensing. If the licenses are compatible merging them is not a bad idea. Nevertheless the blas incompatibility would be an issue.

**Comment by oberbichler** [comment]
I would prefer a solution with one application where
- the `eigen_sparse_lu` is always included as a fast standard solver without binary dependencies
- I have the option to choose from different BLAS/LAPACK libs to speed up all expensive linear algebra operations in Kratos (like large matrix products)
- I can enable specific external solvers like Pardiso, Pastix, SuperLU, ... if they are available

Eigen has this possibility out-of-box (but I think no wrapper for FEAST at the moment). Maybe we can do something similar.

**Pull request merged**